psu/psu.c

```diff
@@ -164,440 +93,550 @@ static inline void set_write_reg_cmd(unsigned char *b, uint32_t addr, uint8_t fo
+	}
+
+	if ((rc = hid_read(dev, b, BUF_SIZE)) < 0 || memcmp(b + 1, &pattern, 4)) {
+		fprintf(stderr, "Failed to receive complete status (rc=%d, status=%02x%02x%02x%02x)\n", rc, b[1], b[2], b[3], b[4]);
+
+		return SCRIPT_ERROR;
 	}
 
-	return 0;
+	fprintf(stderr, " - File has been written correctly.\n");
+
+	return SCRIPT_OK;
 }
 
 
-int write_reg_cmd(hid_device *dev)
+static int sdp_jmpAddr(hid_device *dev, uint32_t addr)
 {
-	long int addr, data, format;
-	char *tok = strtok(NULL, " ");
-	addr = strtol(tok, NULL, 0);
-	tok = strtok(NULL, " ");
-	data = strtol(tok, NULL, 0);
-	tok = strtok(NULL, " ");
-	format = strtol(tok, NULL, 0);
-
-	return sdp_writeRegister(dev, addr, format, data);
+	int rc;
+	unsigned char b[BUF_SIZE] = { 0 };
+
+	fprintf(stderr, " - To the address: %#x\n", addr);
+
+	/* Send write command */
+	b[0] = 1;
+	set_jmp_cmd(b + 1, addr);
+
+	if ((rc = hid_write(dev, b, CMD_SIZE)) < 0) {
+		fprintf(stderr, "Failed to send jump_address command (rc=%d)\n", rc);
+		return rc;
+	}
+
+	/* Receive report 3 */
+	if ((rc = hid_read(dev, b, BUF_SIZE)) < 5) {
+		fprintf(stderr, "Failed to receive HAB mode (rc=%d)\n", rc);
+		return -1;
+	}
+
+	return SCRIPT_OK;
 }
 
 
-int jump_addr_cmd(hid_device *dev)
+static int sdp_errStatus(hid_device *dev)
 {
-	long int addr;
-	char *tok = strtok(NULL, " ");
-	addr = strtol(tok, NULL, 0);
+	unsigned char b[INTERRUPT_SIZE] = { 0 };
+	int rc;
+
+	b[0] = 1;
+	set_status_cmd(b + 1);
+
+	if((rc = hid_write(dev, b, CMD_SIZE)) < 0) {
+		fprintf(stderr, "Failed to send status command (rc=%d)\n", rc);
+		return rc;
+	}
+
+	if((rc = hid_read(dev, b, INTERRUPT_SIZE)) < 5) {
+		fprintf(stderr, "Failed to receive HAB mode (rc=%d)\n", rc);
+		return rc;
+	}
+
+	if((rc = hid_read(dev, b, INTERRUPT_SIZE)) < 0) {
+		fprintf(stderr, "Failed to receive status (rc=%d)\n", rc);
+		return rc;
+	}
 
-	return sdp_jmpAddr(dev, addr);
+	return SCRIPT_OK;
 }
 
 
-int err_status_cmd(hid_device *dev)
+static inline int8_t char_to_hex(char c)
 {
-	return sdp_errStatus(dev);
+	if (c >= '0' && c <= '9') {
+		return c - '0';
+	}
+	else if (c >= 'a' && c <= 'f') {
+		return c - 'a' + 10;
+	}
+	else if (c >= 'A' && c <= 'F') {
+		return c - 'A' + 10;
+	}
+
+	return SCRIPT_ERROR;
 }
 
 
-int parse_byte_string(const char *str, char **out)
+static int parse_byte_string(script_blob_t str, script_blob_t *blob)
 {
-	int res;
-	size_t size = strlen(str);
-	*out = malloc(size * sizeof(*str));
-	char * const out_start = *out;
-	const char * const str_end = str + size;
-	--str;
-
-	while ((++str) < str_end) {
-		if (*str == '\\') {
-			++str;
-			if (*str == '\\') {
-				*((*out)++) = '\\';
-			}
-			else if (*str == 'x') {
-				*(*out)      = (char_to_hex(*(++str)) & 0xf) << 4;
-				*((*out)++) |= (char_to_hex(*(++str)) & 0xf);
-			}
-			else {
-				fprintf(stderr, "Malformed byte string passed\n");
-				free(out_start);
-				*out = NULL;
-				return -1;
-			}
+	if (!(blob->ptr = realloc(blob->ptr, str.end - str.ptr + 1 ))) {
```

Comment:
if `blob->ptr` wasn't NULL, we have memory leak here.

Reply:
`realloc` doesn't free memory on fail

Reply:
That's right, corrected with temporary pointer.